pyscriptjs/src/utils.ts

```diff
@@ -84,4 +86,105 @@ function handleFetchError(e: Error, singleFile: string) {
     showError(errorContent);
 }
 
-export { addClasses, removeClasses, getLastPath, ltrim, htmlDecode, guidGenerator, showError, handleFetchError };
+function readTextFromPath(path: string) {
+    const request = new XMLHttpRequest();
+    request.open("GET", path, false);
+    request.send();
+    const returnValue = request.responseText;
+
+    return returnValue;
+}
+
+function inJest(): boolean {
+    return typeof process === 'object' && process.env.JEST_WORKER_ID !== undefined;
+}
+
+function mergeConfig(inlineConfig: AppConfig, externalConfig: AppConfig): AppConfig {
+    const name = inlineConfig.name || externalConfig.name;
+    const description = inlineConfig.description || externalConfig.description;
+    const version = inlineConfig.version || externalConfig.version;
+    const type = inlineConfig.type || externalConfig.type;
+    const author_name = inlineConfig.author_name || externalConfig.author_name;
+    const author_email = inlineConfig.author_email || externalConfig.author_email;
+    const license = inlineConfig.license || externalConfig.license;
+    const autoclose_loader = (typeof inlineConfig.autoclose_loader !== "undefined") ? inlineConfig.autoclose_loader : externalConfig.autoclose_loader;
+    const runtimes = inlineConfig.runtimes || externalConfig.runtimes;
+    const packages = inlineConfig.packages || externalConfig.packages;
+    const paths = inlineConfig.paths || externalConfig.paths;
+    const plugins = inlineConfig.plugins || externalConfig.plugins;
+    const merged: AppConfig = {
+        name,
+        description,
+        version,
+        type,
+        author_name,
+        author_email,
+        license,
+        autoclose_loader,
+        runtimes,
+        packages,
+        paths,
+        plugins
+    };
+
+    return merged;
+}
+
+function validateConfig(configText: string) {
+    let config: object;
+    try {
+        config = JSON.parse(configText);
+    }
+    catch (err) {
+        console.log(err.message);
+    }
+
+    const finalConfig: AppConfig = {}
+    const allKeys = {
+        "string": ["name", "description", "version", "type", "author_name", "author_email", "license"],
+        "boolean": ["autoclose_loader"],
+        "array": ["runtimes", "packages", "paths", "plugins"]
+    };
+
+    for (const keyType in allKeys)
+    {
+        const keys = allKeys[keyType];
+        keys.forEach(function(item: string){
```

Comment:
we are just validating our configuration basically. All keys are simple -- either strings or arrays -- except `runtimes` which is an array of dicts. So for the `runtimes` key, we check further down one level if the dict it should contain has keys of type `strings`

Reply:
this in house solution is there because we are not allowed to use a schema validator, otherwise it increases overheads

Reply:
But do we need a schema validator..? `</devilsadvocate>`

Code we don't write can't have bugs in it. Also, if someone passes in invalid configuration, they'll soon find out when their code doesn't run as expected. We've got to trust our users to some extent. 😉

Reply:
yeah, so we settled midway, let's not have a library installed just to validate schema, else we have shipping overheads. but still validate it, but we can change our approach if we want.
CC: @tedpatrick

Reply:
But what does such validation give us (apart from more/complicated code to maintain and test)..?

If the user has got it wrong, they'll soon find out (i.e. they don't need us to tell them).

`suggestions = await tedpatrick("opinions please");`

😛

Reply:
The more I think about it, the less I like that we have a validator. Each time we change the schema we're going to have to update this block of code too (along with the other code that uses the config). Why not just throw exceptions in the code that handles the config if we don't find what we're expecting..?

Reply:
Yes, schema validator in next iteration is ok

Reply:
Aha... so @ayirpconda you're saying, no need to have a schema validator (i.e. we can remove this code for now)..?

Reply:
Wait... are we saying that a validator implies us not allowing users to add additional keys to the config? If that's the case I'd be very [very] 👎  . As an App Author I'd like to be able to add whatever extra metadata I want to my app (even if they are not being used by PyScript at all) so that I can use somehow inside my App.

Moreover, if I'm the author of a plugin and I want to allow users to customize my plugin via the AppConfig, how would I do?

Reply:
+100 to what Fabio said.

Reply:
We should validate the important parts and allow any additional data through. 

Are there future properties we should reserve (reserved keys)?

We discussed injecting a "pyscript" object into the <py-config> final data. this would contain version and other system keys.

Reply:
done, we validate stuff important for us, but also allow extra keys and don't do any validation on them.